frame/im-online/src/lib.rs

```diff
@@ -365,156 +410,130 @@ impl<T: Trait> Module<T> {
-		Self::set_worker_status(block_number, true);
+		// acquire lock for that authority at current heartbeat to make sure we don't
+		// send concurrent heartbeats.
+		Self::with_heartbeat_lock(
+			authority_index,
+			heartbeat_at,
+			block_number,
+			|| {
+				let call = prepare_heartbeat()?;
+				debug::info!(
+					target: "imonline",
+					"[index: {:?}] Reporting im-online at block: {:?} (expected: {:?}): {:?}",
+					authority_index,
+					block_number,
+					heartbeat_at,
+					call,
+				);
 
-		Ok(())
-	}
+				T::SubmitTransaction::submit_unsigned(call)
+					.map_err(|_| OffchainErr::SubmitTransaction)?;
 
-	fn compare_and_set_worker_status(
-		gossipping_at: T::BlockNumber,
-		done: bool,
-		curr_worker_status: Option<Vec<u8>>,
-	) -> bool {
-		let enc = WorkerStatus {
-			done,
-			gossipping_at,
-		};
-		sp_io::offchain::local_storage_compare_and_set(
-			StorageKind::PERSISTENT,
-			DB_KEY,
-			curr_worker_status,
-			&enc.encode()
+				Ok(())
+			},
 		)
 	}
 
-	fn set_worker_status(
-		gossipping_at: T::BlockNumber,
-		done: bool,
-	) {
-		let enc = WorkerStatus {
-			done,
-			gossipping_at,
-		};
-		sp_io::offchain::local_storage_set(StorageKind::PERSISTENT, DB_KEY, &enc.encode());
+	fn local_authority_keys() -> impl Iterator<Item=(u32, T::AuthorityId)> {
+		// we run only when a local authority key is configured
+		let authorities = Keys::<T>::get();
+		let mut local_keys = T::AuthorityId::all();
+		local_keys.sort();
+
+		authorities.into_iter()
+			.enumerate()
+			.filter_map(move |(index, authority)| {
+				local_keys.binary_search(&authority)
+					.ok()
+					.map(|location| (index as u32, local_keys[location].clone()))
+			})
 	}
 
-	// Checks if a heartbeat gossip already occurred at this block number.
-	// Returns a tuple of `(current worker status, bool)`, whereby the bool
-	// is true if not yet gossipped.
-	fn check_not_yet_gossipped(
+	fn with_heartbeat_lock<R>(
+		authority_index: u32,
+		heartbeat_at: T::BlockNumber,
 		now: T::BlockNumber,
-		next_gossip: T::BlockNumber,
-	) -> Result<(Option<Vec<u8>>, bool), OffchainErr> {
-		let last_gossip = sp_io::offchain::local_storage_get(StorageKind::PERSISTENT, DB_KEY);
-		match last_gossip {
-			Some(last) => {
-				let worker_status: WorkerStatus<T::BlockNumber> = Decode::decode(&mut &last[..])
-					.map_err(|_| OffchainErr::DecodeWorkerStatus)?;
+		f: impl FnOnce() -> OffchainResult<T, R>,
+	) -> OffchainResult<T, R> {
+		let key = {
+			let mut key = DB_PREFIX.to_vec();
+			key.extend(authority_index.encode());
+			key
+		};
+		let storage = StorageValueRef::persistent(&key);
+		let res = storage.mutate(|status: Option<Option<HeartbeatStatus<T::BlockNumber>>>| {
+			// Check if there is already a lock for that particular block.
+			// This means that the heartbeat has already been sent, and we are just waiting
+			// for it to be included. However if it doesn't get included for INCLUDE_THRESHOLD
+			// we will re-send it.
+			match status {
```

Comment:
this block could be rewritten as a single match evaluating to the `Err` on the first arm and the `Ok` on the second arm. there doesn't seem to be a need for the `return`.